src/script/interpreter.cpp

```diff
@@ -1264,6 +1280,116 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                 }
                 break;
 
+                case OP_VAULT_RECOVER:
+                {
+                    // Only available in Tapscript
+                    if (sigversion == SigVersion::BASE || sigversion == SigVersion::WITNESS_V0) {
+                        return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
+                    }
+
+                    // Stack:
+                    //  - <recovery-sPK-hash>
+                    //  - <recovery-vout-idx>
+                    if (stack.size() < 2) {
+                        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+                    }
+
+                    const valtype& hash_bytes_from_stack = stacktop(-1);
+                    if (hash_bytes_from_stack.size() != 32) {
+                        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+                    }
+                    const uint256 recovery_spk_hash{hash_bytes_from_stack};
+
+                    std::optional<size_t> recovery_vout_idx = GetVoutIdxFromStack(-2);
+                    if (!recovery_vout_idx) {
+                        return set_error(serror, SCRIPT_ERR_VAULT_BAD_VOUT_IDX);
+                    }
+
+                    if (const auto& err = checker.CheckVaultRecover(
+                            *recovery_vout_idx, execdata, recovery_spk_hash, script, flags)) {
+                        return set_error(serror, *err);
+                    }
+
+                    popstack(stack);
+                    popstack(stack);
+                    stack.push_back(vchTrue);
+                }
+                break;
+
+                case OP_VAULT:
+                {
+                    // Only available in Tapscript
+                    if (sigversion == SigVersion::BASE || sigversion == SigVersion::WITNESS_V0) {
+                        return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
+                    }
+
+                    // Stack:
+                    //  - <flu-script>
+                    //  - <n-pushes>
+                    //  [ n items ... ]
+                    //  - <trigger-vout-idx>
+                    //  - <revault-vout-idx>
+                    //  - <revault-amount>
+                    if (stack.size() < 5) {
+                        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+                    }
+
+                    const valtype& flu_script_data{stacktop(-1)};
+                    CScript flu_script_body{flu_script_data.begin(), flu_script_data.end()};
+                    popstack(stack);
+
+                    const CScriptNum n_pushes_data{stacktop(-1), fRequireMinimal};
+                    const size_t n_pushes = static_cast<size_t>(n_pushes_data.GetInt64());
```

Comment:
Is there an indirect reasoning here to why this cast is safe? I would expect a check of `n n_pushes >= 0` here?

Reply:
Good point about the sign check. We know that with default construction arguments, `CScriptNum` will only permit a 4 byte number (`nMaxNumSize`), so from a capacity standpoint this is okay. But checking explicitly for non-negativeness is a good idea.

Reply:
Fixed.